### PR TITLE
[RocksDb] Tables consolidation

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[RocksDb] Tables consolidation

This is one step in a series of changes to move to a one CF per partition-id. The change consolidates all partitions into a single CF `data-unpartitioned`.

Notes:
- This introduces a fixed key prefix to partition CF to improve bloom filter lookups. This change require some subtle handling of scans that spans multiple prefixes, this is handled and is covered by existing test cases.
- This adds the ability for a table to have multiple KeyKinds attached. Each TableKey has an associated kind, and the table (TableKind) has a set of KeyKinds associated.
- This is designed to allow future separation of tables, or even individual key kinds to separate CFs.
- Some values are intentionally passed but not used for (a) clarity at call site and (b) to help with follow-up refactoring.

Side note: cluster marker tests fail sporadically, this adds a flush after each temp file write in an attempt to fix them.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1466).
* #1476
* #1475
* #1474
* __->__ #1466

### CODE REVIEW

File: crates/storage-rocksdb/src/lib.rs [hunk 1]
```diff
@@ -275,40 +289,78 @@ impl RocksDBStorage {
...
         scan: TableScan<K>,
     ) -> DBRawIteratorWithThreadMode<'_, DB> {
         let scan: PhysicalScan = scan.into();
         match scan {
-            PhysicalScan::Prefix(table, prefix) => {
-                let mut it = self.prefix_iterator(table, prefix.clone());
+            PhysicalScan::Prefix(table, key_kind, partition_id, prefix) => {
+                assert!(table.has_key_kind(&prefix));
+                let mut it = self.prefix_iterator(table, key_kind, partition_id, prefix.clone());
                 it.seek(prefix);
```
Comment (tillrohrmann) [inline crates/storage-rocksdb/src/lib.rs:335]: nit: Should the seek become part of the `*_iterator` methods to fully set up the iterator?

File: crates/storage-rocksdb/src/lib.rs [hunk 2]
```diff
@@ -77,40 +88,27 @@ pub enum TableScanIterationDecision<R> {
...
-        Timers => TIMERS_TABLE_NAME,
-        Journal => JOURNAL_TABLE_NAME,
-        Idempotency => IDEMPOTENCY_TABLE_NAME,
-    }
+const fn cf_name(_kind: TableKind) -> &'static str {
+    PARTITION_CF
 }
 
 #[derive(Debug, Copy, Clone, Eq, PartialEq)]
 pub enum TableKind {
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/storage-rocksdb/src/lib.rs:96]: Or is this one of the follow-ups?

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @AhmedSoliman. The changes look good to me. I left a few minor comments. +1 for merging.

File: crates/storage-rocksdb/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/storage-rocksdb/src/lib.rs:96]: I think it's still a useful abstraction, it's now possible to have multiple keys per table, but we might revisit this at some point and clean it up. I'd say that I wouldn't remove it prematurely and only do that when it becomes a hindrance.

File: crates/storage-rocksdb/src/scan.rs [hunk 3]
```diff
@@ -8,67 +8,91 @@
...
+    SinglePartitionKeyPrefix(PartitionKey, K),
+    /// Inclusive Key Range in a single partition.
+    KeyRangeInclusiveInSinglePartition(PartitionId, K, K),
 }
 
 pub(crate) enum PhysicalScan {
-    Prefix(TableKind, BytesMut),
-    RangeExclusive(TableKind, BytesMut, BytesMut),
-    RangeOpen(TableKind, BytesMut),
+    Prefix(TableKind, KeyKind, PartitionId, BytesMut),
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/storage-rocksdb/src/scan.rs:38]: Yes. I'll remove PartitionId actually from here. In a follow up, I plan to make PartitionId and PartitionKey new types to make sure we don't confuse them.

File: crates/storage-rocksdb/src/lib.rs [hunk 4]
```diff
@@ -275,40 +289,78 @@ impl RocksDBStorage {
...
 
         self.db.raw_iterator_cf_opt(&table, opts)
     }
 
-    fn range_iterator(&self, table: TableKind, range: impl rocksdb::IterateBounds) -> DBIterator {
+    fn range_iterator(
+        &self,
+        table: TableKind,
+        _key: KeyKind,
+        maybe_partition_id: Option<PartitionId>,
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/storage-rocksdb/src/lib.rs:311]: I'll add an enum `ScanMode` to make this clear.

File: crates/node/src/cluster_marker.rs [hunk 5]
```diff
@@ -263,6 +263,7 @@ mod tests {
             file.path(),
         )
         .unwrap();
+        file.flush()?;
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/node/src/cluster_marker.rs:267]: Good call. Added that too.

**Pull request merged**